stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/experimental/ws/StreamWebSocket.kt

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.client.socket.experimental.ws
+
+import io.getstream.chat.android.client.errors.ChatError
+import io.getstream.chat.android.client.errors.ChatErrorCode
+import io.getstream.chat.android.client.errors.ChatNetworkError
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.parser.ChatParser
+import io.getstream.chat.android.client.socket.SocketErrorMessage
+import io.getstream.chat.android.client.utils.map
+import io.getstream.chat.android.client.utils.recover
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import okhttp3.Response
+import okhttp3.WebSocket
+import okhttp3.WebSocketListener
+
+private const val EVENTS_BUFFER_SIZE = 100
+private const val CLOSE_SOCKET_CODE = 1000
+private const val CLOSE_SOCKET_REASON = "Connection close by client"
+internal class StreamWebSocket(
+    private val parser: ChatParser,
+    socketCreator: (WebSocketListener) -> WebSocket,
```

Comment:
The Okhttp `WebSocket` constructor requires injecting the listener by the constructor. Because we don't want to have the listener injected outside of our `StreamWebSocket` implementation, a lambda to create the `WebSocket` is required, with this lambda we are able to have an instance of `WebSocket` that use our internal `WebSocketListener`